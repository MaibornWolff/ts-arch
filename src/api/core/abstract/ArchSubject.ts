export abstract class ArchSubject {
  constructor(private name: string)

  public getName(): string {
    return this.name
  }
}
