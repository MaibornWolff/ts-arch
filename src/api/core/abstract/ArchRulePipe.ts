import { ArchSubject } from './ArchSubject'

export abstract class ArchRulePipe {
  constructor(private input: ArchRulePipe) {}

  public filterSubjects(subjects: ArchSubject[]): ArchSubject[] {
    return this.input ? this.input.filterSubjects(subjects) : subjects
  }

  public hasNotModifier(): boolean {
    return this.input ? this.input.hasNotModifier() : false
  }
}
