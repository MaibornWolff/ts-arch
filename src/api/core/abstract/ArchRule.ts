import { ArchSubject } from "./ArchSubject"
import { ArchRulePipe } from "./ArchRulePipe"
import { ArchProject } from "../ArchProject"
import { ArchResult } from "../ArchResult"

export abstract class ArchRule {
	private result: ArchResult

	constructor(public input: ArchRulePipe) {
		this.result = new ArchResult()
	}

	abstract buildResult(subjects: ArchSubject[], hasNotModifier: boolean): ArchResult

	public check(subjects: ArchSubject[]): boolean {
		this.result = this.buildResult(subjects, this.input.hasNotModifier())
		return this.result.hasRulePassed()
	}

	public checkProject(project: ArchProject): boolean {
		return this.check(project.getSubjects())
	}

	public getResult(): ArchResult {
		return this.result
	}
}
