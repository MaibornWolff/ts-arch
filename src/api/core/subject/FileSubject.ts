import { ArchSubject } from '../abstract/ArchSubject'

export class FileSubject extends ArchSubject {
  constructor(name: string, private path: string) {
    super(name)
  }

  public getPath(): string {
    return this.path
  }
}
