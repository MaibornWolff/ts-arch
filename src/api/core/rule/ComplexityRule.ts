import { ArchRule } from "../abstract/ArchRule"
import { ArchRulePipe } from "../abstract/ArchRulePipe"
import { ArchSubject } from "../abstract/ArchSubject"
import { FileSubject } from "../subject/FileSubject"
import { readFileSync } from "fs"
import { createSourceFile, ScriptTarget, Node, SyntaxKind } from "typescript"

export class ComplexityRule extends ArchRule {
	constructor(input: ArchRulePipe, private value: number) {
		super(input)
	}

	public checkCondition(subjects: ArchSubject[]): boolean {
		// TODO how can we manifest types in our pipelines
		let passed = true
		subjects
			.filter(s => s instanceof FileSubject)
			.map(s => s as FileSubject)
			.forEach(s => {
				const mcc = this.getMcc(s)
				if (mcc >= this.value) {
					passed = false
				}
			})
		return passed // TODO return error objects
	}

	// TODO move official typescript parser to Project parser and kick package typescript-parser
	// TODO another option is to use lazy loading like now and then use a cache for already loaded ast's
	public getMcc(subject: FileSubject): number {
		const sourceFile = createSourceFile(
			subject.getName(),
			readFileSync(subject.getPath() + "/" + subject.getName()).toString(),
			ScriptTarget.ESNext,
			false
		)
		return this.analyze(sourceFile)
	}

	private analyze(node: Node) {
		let mcc = 0
		switch (node.kind) {
			case SyntaxKind.CaseClause:
			case SyntaxKind.DefaultClause:
			case SyntaxKind.IfStatement:
			case SyntaxKind.VariableDeclaration:
			case SyntaxKind.CallExpression:
				mcc++
				break
		}
		node.forEachChild(c => {
			mcc += this.analyze(c)
		})
		return mcc
	}
}
