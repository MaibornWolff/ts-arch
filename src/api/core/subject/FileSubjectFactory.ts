import { FileSubject } from "./FileSubject"
import { SourceFile, createSourceFile, ScriptTarget } from "typescript"
import * as path from "path"
import { readFileSync } from "fs"
export class FileSubjectFactory {
	public static buildFromPath(location: string): FileSubject {
		let sf = createSourceFile(location, readFileSync(location).toString(), ScriptTarget.ESNext)
		const name = path.basename(sf.fileName)
		const dir = path.dirname(sf.fileName)
		return new FileSubject(name, dir, sf)
	}

	public static buildFromSourceFile(source: SourceFile): FileSubject {
		const name = path.basename(source.fileName)
		const dir = path.dirname(source.fileName)
		return new FileSubject(name, dir, source)
	}
}
