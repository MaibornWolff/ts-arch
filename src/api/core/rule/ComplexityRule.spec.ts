import { ArchRulePipe } from "../abstract/ArchRulePipe"
import { ComplexityRule } from "./ComplexityRule"
import { FileSubject } from "../subject/FileSubject"

describe("complexity rule", () => {
	let input
	let rule: ComplexityRule

	beforeEach(() => {
		input = jest.fn<ArchRulePipe>(() => {
			return {
				filterSubjects: subjects => subjects,
				hasNotModifier: () => false
			}
		})()
	})

	it("should calculate correct mcc for A.ts", async () => {
		rule = new ComplexityRule(input, 20)
		const subject = new FileSubject("A.ts", __dirname + "/complexitySamples")
		expect(rule.getMcc(subject)).toBe(13)
	})

	it("should pass mcc<20 rule for A.ts", async () => {
		rule = new ComplexityRule(input, 20)
		const subject = new FileSubject("A.ts", __dirname + "/complexitySamples")
		expect(rule.check([subject])).toBe(true)
	})

	it("should not pass mcc<10 rule for A.ts", async () => {
		rule = new ComplexityRule(input, 10)
		const subject = new FileSubject("A.ts", __dirname + "/complexitySamples")
		expect(rule.check([subject])).toBe(false)
	})

	it("should generate correct results for not negated rule when given A.ts and B.ts", async () => {
		rule = new ComplexityRule(input, 5)
		const a = new FileSubject("A.ts", __dirname + "/complexitySamples")
		const b = new FileSubject("B.ts", __dirname + "/complexitySamples")
		expect(rule.check([a, b])).toBe(false)
		expect(rule.getResult()).toMatchSnapshot()
	})

	it("should generate correct results for negated rule when given A.ts and B.ts", async () => {
		rule = new ComplexityRule(input, 5)
		input.hasNotModifier = () => true
		const a = new FileSubject("A.ts", __dirname + "/complexitySamples")
		const b = new FileSubject("B.ts", __dirname + "/complexitySamples")
		expect(rule.check([a, b])).toBe(false)
		expect(rule.getResult()).toMatchSnapshot()
	})
})
