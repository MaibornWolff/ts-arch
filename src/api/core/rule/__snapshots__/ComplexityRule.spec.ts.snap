// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complexity rule should generate correct results for negated rule when given A.ts and B.ts 1`] = `
ArchResult {
  "hasForceFailed": false,
  "hasForcePassed": false,
  "results": Array [
    Object {
      "info": "A.ts has MCC of 13.",
      "pass": true,
      "subject": FileSubject {
        "dir": "C:\\\\Users\\\\AdrianE\\\\Desktop\\\\ts-arch\\\\src\\\\api\\\\core\\\\rule/complexitySamples",
        "name": "A.ts",
      },
    },
    Object {
      "info": "B.ts has MCC of 2.",
      "pass": false,
      "subject": FileSubject {
        "dir": "C:\\\\Users\\\\AdrianE\\\\Desktop\\\\ts-arch\\\\src\\\\api\\\\core\\\\rule/complexitySamples",
        "name": "B.ts",
      },
    },
  ],
}
`;

exports[`complexity rule should generate correct results for not negated rule when given A.ts and B.ts 1`] = `
ArchResult {
  "hasForceFailed": false,
  "hasForcePassed": false,
  "results": Array [
    Object {
      "info": "A.ts has MCC of 13.",
      "pass": false,
      "subject": FileSubject {
        "dir": "C:\\\\Users\\\\AdrianE\\\\Desktop\\\\ts-arch\\\\src\\\\api\\\\core\\\\rule/complexitySamples",
        "name": "A.ts",
      },
    },
    Object {
      "info": "B.ts has MCC of 2.",
      "pass": true,
      "subject": FileSubject {
        "dir": "C:\\\\Users\\\\AdrianE\\\\Desktop\\\\ts-arch\\\\src\\\\api\\\\core\\\\rule/complexitySamples",
        "name": "B.ts",
      },
    },
  ],
}
`;
